lcls_tools/common/measurements/screen_beam_profile_measurement.py

```diff
@@ -0,0 +1,99 @@
+from lcls_tools.common.devices.screen import Screen
+from lcls_tools.common.image.processing import ImageProcessor
+from lcls_tools.common.data.fit.projection import ProjectionFit
+from lcls_tools.common.measurements.measurement import Measurement
+import numpy as np
+
+
+class ScreenBeamProfileMeasurement(Measurement):
+    """
+    ScreenBeamProfileMeasurement class that allows for background subtraction and roi cropping
+    ------------------------
+    Arguments:
+    name: str (name of measurement default is beam_profile),
+    device: Screen (device that will be performing the measurement),
+    image_processor: ImageProcessor ()
+    fitting_tool: ProjectionFit ()
+    fit_profile: bool = True ()
+    ------------------------
+    Methods:
+    single_measure: measures device and returns raw and processed image
+    measure: does multiple measurements and has an option to fit the image profiles
+    """
+
+    name: str = "beam_profile"
+    device: Screen
+    image_processor: ImageProcessor
+    fitting_tool: ProjectionFit
+    fit_profile: bool = True
+    # charge_window: Optional[ChargeWindow] = None
+
+    def measure(self, n_shots: int = 1) -> dict:
+        """
+        Measurement function that takes in n_shots as argumentment, where n_shots is the number of
+        image profiles (is this what I want to call it?)
+        we would like to measure. Invokes single_measure per shot
+        Then if fit_profile = True, fits the x and y projections of each image using the provided
+        fitting_tool. Results are stored in list of dictionaries where each element of the list
+        is results for the image fitting process stored as a dictionary.
+        ----
+        Currently under work, what do we want to do with the images?
+        Needs dump_controller
+
+        """
+        images = []
+        while len(images) < n_shots:
+            measurement = self.single_measure()
+            if len(measurement):
+                images += [measurement]
+
+        results = {"images": images}
+        if self.fit_profile:
+            final_results = []
+            for i, ele in enumerate(images):
+                temp = {}
+                temp["raw_image"] = ele["raw_image"]
+                temp["processed_image"] = ele["processed_image"]
+
+                projection_x = np.array(np.sum(ele["processed_image"], axis=0))
+                projection_y = np.array(np.sum(ele["processed_image"], axis=1))
+
+                for key, param in self.fitting_tool.fit_projection(
```

Comment:
I don't love this syntax.

Reply:
I think there's an impedance mismatch in the way the function is defined and the implementation, and the extra code for stuff that isn't yet implemented or doesn't actually happen (single_measure returning an empty dict).

Below I changed the function definition to yield a dict per image. An iterator in this case has a number of advantages, e.g. memory efficiency:

```py
    def take_and_fit_images(self, n_shots: int = 1) -> dict:
        """Get `n_shots` from the screen and fit the processed images

        Args:
            n_shots (int): number of valid images to take [1]
        Yields:
            dict: `raw_image`, `processed_image`, and `fits`
        """

        def _fit(processed, axis):
            return self.fitting_tool_fit.fit_projection(
                np.array(np.sum(processed, axis=axis))
            )

        for _ in range(n_shots):
            rv = self.single_measure()
            rv["fits"] = dict(
                {k: _fit(rv["processed_image"], v) for k, v in (("x", 0), ("y", 1))}
            )
            yield rv
```

If the shots need to be all taken all at once (without fitting in between each shot), then the higher level code should pass in the shots (per @eloiseyang's suggestion) controlling the speed, and this function could remain an iterator and would be even simpler: